### PR TITLE
Distance sensor downward

### PR DESCRIPTION
**Pull request by baumanta**
This PR enables the use of a downward facing distance sensor for low flight altitudes. The local planner sees and treats the ground as an obstacle if it lies inside the field of view. This is a problem when flying low using obstacle avoidance or collision prevention. When using the obstacle avoidance the drone gets trapped, because even if it yaws, it will see ground in all directions. When using collision prevention, the forward movement is prevented at low altitudes. This is not a desired behavior.

The distance sensor input is read in and the pointcloud is cropped such that we are sure, the ground is not inside the FOV (except where there are steep slopes, but there we might want to prevent getting closer). There are a lot of hardcoded numbers. They come from the tuning (realsense noise, etc.). In the future we will aim at a more generic solution

### CODE REVIEW

File: local_planner/launch/rs_depthcloud.launch [hunk 1]
```diff
@@ -45,14 +45,14 @@ Processing enabled by this node:
   <!-- Camera device specific arguments -->
 
   <arg name="serial_no"           default=""/>
-  <arg name="json_file_path"      default=""/>
+  <arg name="json_file_path"      default="/usr/auterion_ws/src/avoidance/local_planner/resource/stereo_calib.json"/>
```
Comment (mrivi) [inline local_planner/launch/rs_depthcloud.launch]: not sure if this belongs in the open source project

File: sim/models/iris_depth_camera_3/iris_depth_camera.sdf [hunk 2]
```diff
@@ -148,7 +122,7 @@
           <cameraInfoTopicName>rgb/camera_info</cameraInfoTopicName>
           <depthImageTopicName>depth/image_raw</depthImageTopicName>
           <depthImageCameraInfoTopicName>depth/camera_info</depthImageCameraInfoTopicName>
-          <pointCloudTopicName>depth/points</pointCloudTopicName>
+          <pointCloudTopicName>depth_registered/points</pointCloudTopicName>
```
Comment (mrivi) [inline sim/models/iris_depth_camera_3/iris_depth_camera.sdf]: why did you change the namespace? This brakes https://github.com/PX4/avoidance/blob/master/local_planner/src/nodes/local_planner_node.cpp#L161-L164 which expects `*/points` and `*/camera_info`to be in the same namespace.

File: local_planner/launch/rs_depthcloud.launch (hunk 1 above)
Comment (baumanta) [reply local_planner/launch/rs_depthcloud.launch]: you are right it should be relative path to the directory inside the local_planner package. Some how it did not work with relative paths at first try. But I will try to fix it

File: sim/models/iris_depth_camera_3/iris_depth_camera.sdf (hunk 2 above)
Comment (baumanta) [reply sim/models/iris_depth_camera_3/iris_depth_camera.sdf]: ok, sure. I changed it because one of the launch files would not work with those topic names. But now I changed it the other way round such that the launch file take depth/points as an input

File: local_planner/src/nodes/local_planner_node.cpp [hunk 3]
```diff
@@ -228,6 +233,16 @@ void LocalPlannerNode::updatePlannerInfo() {
     local_planner_.setGoal(goal_msg_.pose.position);
     new_goal_ = false;
   }
+
+  // update ground distance
+  if (ros::Time::now() - ground_distance_msg_.header.stamp <
+      ros::Duration(0.5)) {
+    local_planner_.ground_distance_ = ground_distance_msg_.bottom_clearance;
+  } else {
+    local_planner_.ground_distance_ = 2.0;  // in case where no range data is
```
Comment (nicovanduijn) [inline local_planner/src/nodes/local_planner_node.cpp]: So we consider 2.0 meters "close"? I'd call 2 cm close. Am I correctly assuming measurement in meters? I'd prefer if the variable name reflected the units...

File: local_planner/src/nodes/star_planner.h [hunk 4]
```diff
@@ -106,7 +107,7 @@ class StarPlanner {
                      double height_change_cost_param_adapted,
                      double height_change_cost_param);
   void setPose(const geometry_msgs::PoseStamped& pose);
-  void setBoxSize(const Box& histogram_box);
+  void setBoxSize(const Box& histogram_box, const double ground_distance);
```
Comment (nicovanduijn) [inline local_planner/src/nodes/star_planner.h]: Missing &.
Passing by const value makes no sense, const & or by value.

File: local_planner/src/nodes/local_planner.cpp [hunk 5]
```diff
@@ -77,7 +75,7 @@ void LocalPlanner::dynamicReconfigureSetParams(
 }
 
 void LocalPlanner::setVelocity() {
-  const auto& p = curr_vel_.twist.linear;
+  const auto &p = curr_vel_.twist.linear;
```
Comment (jkflying) [inline local_planner/src/nodes/local_planner.cpp]: Did the auto-format do this?

File: local_planner/src/nodes/star_planner.h (hunk 4 above)
Comment (jkflying) [reply local_planner/src/nodes/star_planner.h]: For a double, just by value makes the most sense I think.

File: local_planner/src/nodes/box.h [hunk 6]
```diff
@@ -10,19 +10,19 @@ class Box {
...
 
-  void setBoxLimits(const geometry_msgs::Point& pos);
+  void setBoxLimits(const geometry_msgs::Point& pos,
+                    const double ground_distance);
   bool isPointWithinBox(const double& x, const double& y, const double& z);
 
   double radius_;
-  double zsize_up_;
-  double zsize_down_;
+  double box_dist_to_ground_ = 2.0;
```
[the remark this answers is not in the document]
Comment (baumanta) [reply local_planner/src/nodes/box.h]: they are needed for the visualization. And I found it to be over kill to write a setter/getter function

**Comment by mrivi** [comment]
@baumanta what's the state of this PR? Flight tests? In any case it needs rebasing on master

**Comment by baumanta** [comment]
Yes, it is somewhat on hold. We had some terrible flight tests this week. I'm pretty sure it is unrelated to those changes but it is on hold until we can show clearly that this is not what broke the stuff

**Comment by jkflying** [comment]
Tested and working on a single-Realsense frame.

**Comment by nicovanduijn** [comment]
@jkflying and I just had a brief discussion, concerning this. Currently, the input is the `/mavros/altitude` topic, but it should really be `/mavros/distance_sensor`. This would start to matter @AnnaDaiZH creates a new custom mavlink stream
